### PR TITLE
Add some comments on FromBuiltin/ToBuiltin/Lift

### PR DESCRIPTION
**Pull request by michaelpj**
Hopefully this helps a bit.

### CODE REVIEW

File: plutus-tx/src/PlutusTx/Builtins/Class.hs [hunk 1]
```diff
@@ -220,4 +165,80 @@ instance ToBuiltin BLS12_381.Pairing.MlResult BuiltinBLS12_381_MlResult where
...
 
+This is what the ToBuiltin and FromBuiltin classes do. They let us write wrappers
+for builtins relatively consistently by just calling toBuiltin on their arguments
+and fromBuiltin on the result. They shouldn't really be used otherwise.
+
+To keep the consistency we define dummy instances that are just the identity function
+for types that don't have a separate "normall Haskell" version. For example:
+integer. Integer in Plutus Tx user code _is_ the opaque builtin type, we don't
+expose a different one. Essentially: if it's a datatype then there's a substantive
+conversion.
```
Comment (effectfully) [inline plutus-tx/src/PlutusTx/Builtins/Class.hs:188]: This sounds like we have two cases:

1. `Bool` vs `BuiltinBool` -- Plutus Tx type _vs_ built-in type, both appear in the `ToBuiltin`/`FromBuiltin` instance
2. `Integer` vs `BuiltinInteger` -- Plutus Tx type _and_ built-in type, both appear in the `ToBuiltin`/`FromBuiltin` instance

but we actually have a third one:

3. `Data` vs `BuiltinData` -- Plutus Tx type _vs_ built-in type, only the latter appears in the `ToBuiltin`/`FromBuiltin` instance

So this Note doesn't clarify this particular quirk for me.

@zliu41 did provide some explanations in [slack](https://input-output-rnd.slack.com/archives/G01EE4NQ9U0/p1695125397674719?thread_ts=1695059755.064939&cid=G01EE4NQ9U0), so could you please incorporate those if you agree with them?

File: plutus-tx/src/PlutusTx/Builtins/Class.hs [hunk 2]
```diff
@@ -220,4 +165,80 @@ instance ToBuiltin BLS12_381.Pairing.MlResult BuiltinBLS12_381_MlResult where
...
+We could use a type family here to get the builtin representation of a type. After all, it's
+entirely determined by the Haskell type.
+
+However, this is harder for the plugin to deal with. It's okay to have a type variable
+for the representation type that needs to be instantiated later, but it's *not* okay to
+have an irreducible type application on a type variable. So fundeps are much nicer here.
+-}
+
+{- Note [Strict conversions to/from unit]
+Converting to/from unit *should* be straightforward: just ``const ()`.`
```
Comment (effectfully) [inline plutus-tx/src/PlutusTx/Builtins/Class.hs:201]:
```suggestion
Converting to/from unit *should* be straightforward: just `const ()`.
```

File: plutus-tx/src/PlutusTx/Lift/Class.hs [hunk 3]
```diff
@@ -135,56 +135,88 @@ instance (TypeError ('Text "Int is not supported, use Integer instead"))
...
+to put inside the constant.
+
+Fortunately, we have To/FromBuiltin, which happen to do what we want.
+See Note [Builtin types and their Haskell versions].
+This is arguably slightly an abuse: the versions of the types that we want in
+Plutus Tx source code and the versions that we use as the implementations of
+the builtin types in the universe could be different. But in practice they
+aren't. So we can write fairly straightforward instances for most types.
+
+Similarly, for Typeable we may have to use a different type from the opaque one.
```
Comment (effectfully) [inline plutus-tx/src/PlutusTx/Lift/Class.hs:221]: Thank you, that makes sense to me. But it still doesn't address the inconsistencies, I described those in the Jira [ticket](https://input-output.atlassian.net/browse/PLT-6538?focusedCommentId=213564), but I'll duplicate them here as well.

#5547 adds a bunch of instances and I wonder if that's what we want. In particular:

```diff
+ instance uni `PLC.HasTypeLevel` () => Typeable uni BuiltinUnit where
+     typeRep _proxyUnit = typeRepBuiltin (Proxy @())
+
+ instance uni `PLC.HasTermLevel` () => Lift uni BuiltinUnit where
+     lift = liftBuiltin . fromBuiltin
+
+ instance uni `PLC.HasTypeLevel` Bool => Typeable uni BuiltinBool where
+     typeRep _proxyBool = typeRepBuiltin (Proxy @Bool)
+
+ instance uni `PLC.HasTermLevel` Bool => Lift uni BuiltinBool where
+     lift = liftBuiltin . fromBuiltin
+
+ instance uni `PLC.HasTypeLevel` [] => Typeable uni BuiltinList where
+     typeRep _proxyBuiltinList = typeRepBuiltin (Proxy @[])
+
+ instance (FromBuiltin arep a, uni `PLC.HasTermLevel` [a]) => Lift uni (BuiltinList arep) where
+     lift = liftBuiltin . fromBuiltin
+
+ instance uni `PLC.HasTypeLevel` (,) => Typeable uni BuiltinPair where
+     typeRep _proxyBuiltinPair = typeRepBuiltin (Proxy @(,))
+
+ instance (FromBuiltin arep a, FromBuiltin brep b, uni `PLC.HasTermLevel` (a, b)) =>
+         Lift uni (BuiltinPair arep brep) where
+     lift = liftBuiltin . fromBuiltin
```

Also, why do we have

```haskell
instance uni `PLC.HasTermLevel` BS.ByteString => Lift uni BS.ByteString where
    lift = liftBuiltin
```

? It appears to be the only instance for a non-`Builtin*` type that isn't equal to its built-in version (i.e. `BuiltinInteger`/`Integer`).

**Review by effectfully — COMMENTED** [review]
This does help me understand the setup better, but it doesn't answer my questions regarding various inconsistencies, I replicated those in the review comments.

File: plutus-tx/src/PlutusTx/Lift/Class.hs (hunk 3 above)
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Lift/Class.hs:221]: I think those instances are indeed correct. And the one for ByteString looks wrong :scream:

File: plutus-tx/src/PlutusTx/Builtins/Class.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Builtins/Class.hs:188]: Hmm, I agree that the `Data` case is odd. I am actually not sure why it's like that. I'm going to try changing it and see what breaks.

File: plutus-tx/src/PlutusTx/Builtins/Class.hs [hunk 4]
```diff
@@ -220,4 +165,80 @@ instance ToBuiltin BLS12_381.Pairing.MlResult BuiltinBLS12_381_MlResult where
...
+case on unit (`chooseUnit`) as a builtin. But then it all works okay.
+-}
+
+{- Note [noinline hack]
+For some functions we have two conflicting desires:
+- We want to have the unfolding available for the plugin.
+- We don't want the function to *actually* get inlined before the plugin runs, since we rely
+on being able to see the original function for some reason.
+
+'INLINABLE' achieves the first, but may cause the function to be inlined too soon.
```
[the remark this answers is not in the document]
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Builtins/Class.hs:217]: Yeah, although the current hack has worked okay for some time.

File: plutus-tx/src/PlutusTx/Lift/Class.hs (hunk 3 above)
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Lift/Class.hs:244]: We needed this for a derived lift instance for Data, which we also shouldn't have. I removed both and it's fine.

File: plutus-tx/src/PlutusTx/Builtins/Class.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-tx/src/PlutusTx/Builtins/Class.hs:188]: It's actually quite odd, I'm trying to pin down how it should work.

**Pull request merged**